lib/srv/usermgmt.go

```diff
@@ -389,22 +316,44 @@ func (u *HostUserManagement) UpsertUser(name string, ui *services.HostUsersInfo)
 		}
 
 		return nil
-	})
-	if err != nil {
-		return nil, trace.Wrap(err)
+	}))
+}
+
+// UpsertUser creates a temporary Teleport user in the TeleportServiceGroup
+func (u *HostUserManagement) UpsertUser(name string, ui services.HostUsersInfo) (io.Closer, error) {
+	var groupErrs []error
+	// cloning to prevent unintended mutation of passed in Groups slice
+	ui.Groups = slices.Clone(ui.Groups)
```

Comment:
Unfortunately yes. In the tests we're passing sub-slices to `userinfo.Groups` which still reference the original slice. `append` tries to be performant and re-use the underlying memory, but doesn't have enough context to know whether an operation is safe and can leak mutations to the original slice. The specific case I noticed was essentially:
```go
allGroups := []string{"foo", "bar"}
addGroups := allGroups[:1]
_ = append(addGroups, "baz")

// allGroups == ["foo", "baz", "bar"]
// addGroups = ["foo", "baz"]
```
Since this is fairly sneaky I opted to prevent it in the function itself rather than use `slices.Clone()` in the tests